### DISCUSSION TITLE
Espressif/Arduino Core library upgrade

### DISCUSSION THREAD

**Discussion by proddy** [emsesp/EMS-ESP32#2481] [Ideas]
Following up from #1804, after 3.7.0 is out look at migrating some of the core libraries and frameworks to see if it improves performance, memory consumption and the firmware binary filesize isn't too big.

- [ ] https://github.com/pioarduino/platform-espressif32 to use Arduino v3.1/IDF v5.3 (from v2.0.x/IDF v4.4.x)
- [ ] Web Server to https://github.com/mathieucarbou/ESPAsyncWebServer and then PsychicHTTP with SSL when it's mature
- [ ] MQTT client espmqttclient -> https://github.com/mlesniew/PicoMQTT (no more AsyncTCP)

**Comment by MichaelDvP** [comment]
The factory partition for 16M is already nearly full. When creating it the idea was to put only a small loader in, not the full emsesp as now. So better use symmetrical partition sizes (3.6.5 have different partitions, so it fits in the braking change of 3.7.0 if we change now). Should also be used in emsesp-flasher.
Suggestion:
```
# Name,   Type, SubType, Offset,   Size,    Flags
nvs,      data, nvs,     0x9000,   0x005000,
otadata,  data, ota,     ,         0x002000,
boot,     app,  factory, ,         0x480000,
app0,     app,  ota_0,   ,         0x490000,
app1,     app,  ota_1,   ,         0x490000,
nvs1,     data, nvs,     ,         0x040000,
spiffs,   data, spiffs,  ,         0x200000,
coredump, data, coredump,,         0x010000,
```

**Comment by proddy** [comment]
Are you suggesting we keep boot smaller and put in https://github.com/emsesp/EMS-ESP-Loader ? I'm fine with that. Now is the time for breaking changes.

It would mean changing a lot though, from platform.io, the bbqkees flash too, the Flash Tool and the Web Installer. And we probably need to put back in the WebUI the option to restart another partition. 

PIO always uploads to `boot` but doesn't adjust the partition info, so sometimes, when you restart after and esptool.py, it will go back to the old firmware version. I think this is a bug in platformio that Jason2866 just fixed in pioarduino.

I've updated the notes I keep on the partitions here. Can you review it please? 
https://github.com/emsesp/bbqkees_flash/blob/main/README.md#partition-explanation

(not to others; this is private repo so you may not have access)

**Comment by MichaelDvP** [comment]
> Are you suggesting we keep boot smaller and put in https://github.com/emsesp/EMS-ESP-Loader ?

No, i suggust  to use normal emsesp and make boot nearly same size as ota (now boot is 280000, i want to enlarge to 480000).
The loader is only usefull if we ran out of mem in 4M chips.

**Comment by proddy** [comment]
ok. why not make all app sizes the same size? This change will erase the nvs data (nrg's and bbqkees's codes). BBQKees codes should be ok as this is not yet in production. The energy counters, is this going to hurt people? If so, maybe make export to Json in the WebUI and upload. The settings will also go, but they can be backed up. I do think we need to do this asap in 3.7.0. And I really want 3.7.0 out in the next weeks as I'm going to be busy at work for the next 6 months

**Comment by MichaelDvP** [comment]
> ok. why not make all app sizes the same size?

It's nearly same size (64k diff on 4,7M partition), but i don't want to move nvs1/spiffs and keep sizes/bounderies on round values.

**Comment by proddy** [comment]
ah I see. Ok, let's do it.

**Comment by proddy** [comment]
btw I tried the latest pioarduino 3.0.7 and used heap went down from 208->196, max alloc too, and bin size +280KB so I think we'll stick with Arduino 2.x until 3.1 comes out.

**Comment by MichaelDvP** [comment]
With pioarduino the network scan still is broken: "no networks found"

**Comment by proddy** [comment]
> With pioarduino the network scan still is broken: "no networks found"

should be fixed in the 3.1.0 - https://github.com/espressif/arduino-esp32/issues/ 10281 hopefully

more info at https://discord.com/channels/1263397951829708871/1263397951829708874/1287027557677662261

**Comment by proddy** [comment]
We can try Arduino Core 3.1 and ESP-IDF 5.3 with

`platform = https://github.com/pioarduino/platform-espressif32/releases/download/53.03.10-rc2/platform-espressif32.zip`

but will crash without patching NTP and AsyncTCP. See https://github.com/espressif/arduino-esp32/issues/ 10526

**Comment by proddy** [comment]
I started porting over to Arduino Core 3.1 / IDF 5.3 and Pioarduino. I also used the latest AsyncTCP and AsyncWS libraries, which I know work. One nice thing about these new libraries is that they give us back 6-7KB of available heap.

But I'm getting stack dumps and reboots when started. It's hard to trace where or why it's breaking, and I have a feeling it memory memory-related, but I can't figure out why it used to work. If I take the backtrace and use for example:

```
addr2line -pfiaC -e .pio/build/s3_16M_P/firmware.elf 0x4037f57b:0x3fcb36e0 0x4037718f:0x3fcb3710 0x403771c9:0x3fcb3730 0x4037f759:0x3fcb3750
```

it doesn't tell me much. So very hard to debug. I'll keep digging.

@MichaelDvP FYI

It's in my branch at https://github.com/proddy/EMS-ESP32/tree/core_update_2108

**Comment by MichaelDvP** [comment]
First look: you have reduced the arduino loop stack to 4k (default 8k), this is critical low. 
Scheduler stack is a bit low when using http-client, maybe enlarge or disable schedules for testing.
Afaik arduino3 have changed uart handlng, could be a test to enlarge the uart stack size.

**Comment by proddy** [comment]
Thanks Michael! I removed the  `-D ARDUINO_LOOP_STACK_SIZE=4096` and it runs again.

**Comment by proddy** [comment]
now I need to fix the Network scanning and add back the Tabs underline in the WebUI

**Comment by proddy** [comment]
Got it all working on Arduino 3.1/IDF 5.3 except it still takes up 50K more heap and max alloc, so really need to try and get the Hybrid working to remove the unnecessary components.

**Comment by proddy** [comment]
I pushed the PR (https://github.com/emsesp/EMS-ESP32/pull/2328) that changes the src code locations and uses the remote libraries for ESPAsync and ESPAsyncWebServer. In 3.7.2-dev-8.

Only yesterday I got a reboot of my EMS-ESP so hope it's not related. If it happens again, I'll roll back the changes

**Comment by Delta1977** [comment]
I'm now for 16h on dev8 without problems. No unexpected reboot after upgrade.

**Comment by proddy** [comment]
@MichaelDvP how do you feel about trying to get EMS-ESP upgraded to the latest Arduino Core 3.1.10 (IDF 5.3.4)? There is some work to do to reduce the footprint by disabling libraries (like Tasmota does using Hybrid) and changing the WiFi code.

Although my ultimate goal is to eventually get off Arduino and go native IDF. Most of our libraries already support IDF.

I'm worried that if we're still on Core v2, we'll never get off it in the future.

**Reply by MichaelDvP** [reply]
The arduino 3 is still memory hungry. Don't think this is good for builds without psram.
The 16MB+ builds can use the actual pioarduino. It compiles with a lot of warnings and runs with some issues. 
- mqtt works after manual start, but after poweron-reset mqtt stays unconnected until user dis/reenables.
- ntp does not work

WiFi-scan works now, ems-bus and syslog also working.

Memory compare for a standalone E32V2.2 (bus not connected):
```
   "arduino": "ESP32 Arduino v2.0.17",
    "sdk": "v4.4.7-dirty",
    "freeMem": 139,
    "maxAlloc": 99,
    "freeCaps": 4183,
    "usedApp": 2416,
```
```
    "arduino": "ESP32 Arduino v3.3.7",
    "sdk": "v5.5.2-729-g87912cd291",
    "freeMem": 117,
    "maxAlloc": 87,
    "freeCaps": 4163,
    "usedApp": 2616,
```

**Reply by proddy** [reply]
yes, what is needed is to use the hybrid mode when building, and selecting only the espressif libraries we need. See example on https://github.com/pioarduino/platform-espressif32/blob/main/examples/tasmota_platformio_override.ini 

I tried it once then gave up. This is how jason builds their Tasmota libraries like https://github.com/tasmota/platform-espressif32/releases/tag/2026.02.30 and removes SSL for example

**Reply by MichaelDvP** [reply]
Found what is wrong with mqtt/ntp. The `WiFi.onEvent()` is not working, the callback functions are never executed (mqtt, ntp, network, AP services). It was wokring in earlier pioarduino, gives no compile warning, only is not executed.

**Reply by MichaelDvP** [reply]
I've tested the latest tasmota for 4MB chip. `platform = https://github.com/tasmota/platform-espressif32/releases/download/2026.02.30/platform-espressif32.zip`
The code is larger, but still fits in the partitions of 4M chip. Free ram-memory is better!
```
    "version": "3.8.2-dev.2",
    "arduino": "Tasmota Arduino v2.0.17",
    "sdk": "4.4.8.240628",
    "freeMem": 189,
    "maxAlloc": 107,
    "freeCaps": 141,
    "usedApp": 1774,
    "freeApp": 210,
```
```
    "version": "3.8.2-dev.4",
    "arduino": "Tasmota Arduino v3.3.7",
    "sdk": "5.3.4.260127",
    "freeMem": 198,
    "maxAlloc": 108,
    "freeCaps": 147,
    "usedApp": 1901,
    "freeApp": 83,
```

**Reply by proddy** [reply]
nice! I was thinking we could just completely switch to Tasmota's core for all variants. It already includes the BearSSL libraries used by WiFiClientSecure and if we drop that library it will save us 600K flash. For TLS we would need to adjust the espmqttclient code. Is this worth trying?

**Reply by MichaelDvP** [reply]
Still `WiFi.onEvent`does not work, i've checked the arduino 3 examples, they use `Network.onEvent`, but this also does not call the callback in latest pioarduino, So ETH does not work and logging of connection does not work and auto-tx-power does not work. 
For mqtt, ap, ntp we can replace the onEvent by polling in loop, but for networkservice this is a bigger refactoring. 
The pioarduino often changes the api without notice, what's working is changing from release to release. 
When moving to IDF we should stay on arduino 2 and not work on both.

**Reply by proddy** [reply]
I managed to get EMS-ESP working on Tasmota Arduino Core 3.1.10/IDF 5.3.4 using `platform = https://github.com/tasmota/platform-espressif32/releases/download/2026.02.30/platform-espressif32.zip` with TLS by manually adding the mbedtls_ssl library which is specifically excluded from Tasmota as they use their own BearSSL library for ESP8266. Both RAM and Flash are considerably smaller, the heap is still the same. If we can fix the onEvents I think this is a path forward. I'll make a specific branch for this.

**Reply by proddy** [reply]
@MichaelDvP I merged my commits to a new branch called core3 (https://github.com/emsesp/EMS-ESP32/tree/core3)

It uses the Tasmota Arduino 3/IDF 5.3.4 for all builds now and works with TLS. If you could add your changes, then I think this could be a good alternative to get EMS-ESP back on the latest stream. We can reconsider a full IDF port later.

Plus the .onEvents work with Tasmota

**Reply by MichaelDvP** [reply]
Compiles without warnings. 
- LittleFS is changed, formats the filesystem with `"No settings found on filesystem. Using factory settings."`
- `onEvent` does not work, tested using `WiFi.onEvent(` and `Network.onEvent(`, no callback executed for both variants.

I haven't tested TLS connections.

**Reply by proddy** [reply]
thanks for those fixes. What we need to look into is indeed
- the LittleFS can't load previous partition. This could be because the bootloader size may have changed the partition alignment. 
- changing Version.tsx as auto-install is not working because we check for Tasmota in the version 
```js
    if (systemData.arduino_version.startsWith('Tasmota')) {
      setDownloadOnly(true);
    }
```

**Reply by MichaelDvP** [reply]
With arduino3 build in app1 partition the esp always reboots to app0. Via serial console i can `restart app1` from arduino2 build in app0 and it boots app1 with always erasing the FS. Commands `restart`, `restart app1` from arduino3 build always boot app0 partition.
Is the otadata with boot info bigger now and do not fit in otadata partition?

**Reply by proddy** [reply]
The FS is erased becuse of the `LittleFS.begin(true)` failing. I don't know why it can't find the filesystem as it's in the correct location. If I add this line to Core v2 and Core v3 they both show `0x00DF0000` so the flash partition is the same. weird. 

```C
    const esp_partition_t * part = esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_DATA_SPIFFS, NULL);
    if (part != NULL) {
        LOG_DEBUG("Confirmed Offset: 0x%08X\n", part->address);
    } else {
        LOG_DEBUG("No SPIFFS partition found");
    }
```

**Reply by MichaelDvP** [reply]
> The FS is erased becuse of the LittleFS.begin(true)

Yes, this formats the filesystem, so the confg is not found. I've tried with `LittleFS.begin(false)` and this fails.

I've flashed the core3 build to both app partitions and now it boots correctly and does not format filesystem on reboot. Also restart command in app1 starts to app1 again. In console the partitions also show correctly:
```
 Partitions:
  app0: v3.8.2-dev.7 (4672 KB, installed 2026-02-17T10:15:18)
  app1: v3.8.2-dev.7 (4672 KB, installed 2026-02-17T08:43:04) ** active **
  boot: v3.8.2-dev.1 (4608 KB, installed 2026-01-24T10:19:38)
```
but in web status/version the other partitions are not shown for reinstall. 
**Edit:** It was not in developer mode, now the other partitions are shown in web.

**Reply by proddy** [reply]
I'll do some further tests tonight if I have time to see if there is a difference between partitions on 3.8.1 and our core3 version

**Reply by MichaelDvP** [reply]
There is no difference.
If this is the sdkconfig from tasmota https://github.com/pioarduino/platform-espressif32/blob/main/examples/tasmota_platformio_override.ini it differs from sdkconfig.default also in some settings of bootloader and littleFS.
But it's compatible with out arduino2 4/16MB tasmota builds. 
Flashing both app0,app1 partitions (via upload) with core3/tasmota builds works. You can change partitions and filesystem stays on reboot.

**Reply by proddy** [reply]
Ok. What the reason it couldn't find the FS when upgrading from core 2? I only tried using the http uploader

**Reply by MichaelDvP** [reply]
Yes, seems there is something incompatible and core3 formats the partition in `LittleFS.begin(true)` 
Not only after flashing, also if one partition contains core 2 (16MB+), and you reboot to the partitions containing core3.
If you come from core2 tasmota and reboot to core3 the filesystem can be read, no format and files can be read.

Maybe the LittleFS lib has changed or an effect of the SDKconfig tasmota:
```
CONFIG_LITTLEFS_MULTIVERSION=y
# CONFIG_LITTLEFS_DISK_VERSION_MOST_RECENT is not set
# CONFIG_LITTLEFS_DISK_VERSION_2_1 is not set
CONFIG_LITTLEFS_DISK_VERSION_2_0=y
```
vs. sdkconfig from out core2 builds:
```
# CONFIG_LITTLEFS_MULTIVERSION is not set
```

**Reply by proddy** [reply]
Hi @MichaelDvP - I spent some time A/B testing Core 2 and Core 3 to see what changes were made in the Litt;leFS that prevent Core 3 from reading our EMS-ESP filesystem. I got nowhere! 

I still like the idea of switching over to Tasmota. The flash is a lot smaller, and the gains in heap are useful. There can be only a single VFS on the ESP so it's not possible to mount in the core2 style and then save in core3 and have this automatic during the EMS-ESP boot. If we take this path, we would need to instruct users to make a backup of all files (we can add a new Web button for this) and then re-import, but this is tricky for remote users and will brick a lot of devices. Another method is to create a Python script that downloads the FS, formats and sends it back (using mklittlefs and esptool). Can you think of any ideas?

**Reply by MichaelDvP** [reply]
I've also made some tests.
- try hybrid build with sdkconfig.defaults file. No luck. The pioarduino creates files and compile the complete cores, but mix up pathes and fail when trying to include `common.h` from false pathes. 

LitleFS formatting:
- update core2 platform-espressif to  pioarduino core3 formats filesystem
- update core2 platform-espressif to  tasmota core 3  formats filesystem
- update core2 platform-espressif to  tasmota core 2  also formats filesystem
- update core2 tasmota to  tasmota core 3 works without formatting
- update core3 tasmota to  pioarduino core 3 works without formatting
- update core3 pioarduino to  tasmota core 3 works without formatting

Bootloader:
if the app0/1 partitons contains different cores, a reboot always boots the core 2. We have to write core 3 to app0 and app1 to boot the latest partition.

Secure mqtt:
with your tls-implementation i can not connect `test.mosquitto.org:8883` with certificate. Works with current core2, but only after reboot, Disconnect/connect on reconfigure fails. I have tried a different implementation https://github.com/mobizt/ESP_SSLClient, that uses bear-ssl
That works the same way as core 2. Also connect after reboot, but no connect on reconfigure. Don't know if this timing from test.mosquitto.org or a ems-esp issue.
With this SSL-client i also implemented https://github.com/mobizt/ReadyMail as a system/sendmail command. Tested with STARTTLS on my mailbox. SSL-client and mail takes about 50k flash for the tasmota build. (core 2 secureWiFiClient does not support STARTTLS).

For using mklittlefs to restore settings you have to know what is different. I have not found the difference.
But i think moving to tasmota core 3 should count major version to 3.9.0 and we can warn the users about changed filesystem and the needed reconfiguration. I think the bootloader issue is more problematic to explain to users. 
Maybe we can check on booting core 3 if the other partiton contains a core 2 firmware and make the second partition unbootable (overwrite partition header). Don't know if this will work.

**Reply by proddy** [reply]
The system backup and restore is in both versions now (dev-16 and core3), along with a clear message in the WebUI when upgrading to a major version. There's also an FAQ article in the docs that is referenced. So there's not much we can do now. 

Next steps?
 Merge core3 into dev and just see what happens? I guess coordinate with @bbqkees first as his first line support may get a lot of questions.

**Reply by MichaelDvP** [reply]
Imho critical is the strange reset to core 2 partition on a reboot with mixed versions in the two app partitions. 
We need a hint for users to flash clean (with erase) via usb or update twice (both partitions). before changing any setting that requires a reboot.

Also there are some smaller cosmetical issues: The dialogbox for download backup stays open after dwnload andhasto closed by cancel-button. The warning does not appear when updating from 3.8.2-dev.16 to 3.9.0-dev.0.

**Reply by proddy** [reply]
> Imho critical is the strange reset to core 2 partition on a reboot with mixed versions in the two app partitions.

I'm hoping we can solve this. Maybe always boot to 'boot' partition which finds the latest? 

> The dialogbox for download backup stays open after dwnload andhasto closed by cancel-button. 

thought I fixed that?

>  The warning does not appear when updating from 3.8.2-dev.16 to 3.9.0-dev.0

hmm, I'll look into that. Are you drag/dropping the .bin or using the Update button?

**Reply by MichaelDvP** [reply]
> I'm hoping we can solve this. 

Seems to work if we set `esp_ota_set_boot_partition(esp_ota_get_running_partition());` in system_restart Maybe also set this after update in update check.
> thought I fixed that?

Ups, yes, it is fixed, i was not on the latest dev16.
> hmm, I'll look into that. Are you drag/dropping the .bin or using the Update button?

click in the drop field and select in filebox. Same with drag&drop.

**Reply by MichaelDvP** [reply]
I think it's this check, the first check should be latest or skipped (return 0). 
https://github.com/emsesp/EMS-ESP32/blob/cb3c9653ce12cfdb68436ce9c3d88c9c5cb73250/src/web/WebStatusService.cpp#L299-L311

**Reply by proddy** [reply]
@MichaelDvP I was playing around with OLED (see discussion) so now have time to fix some of these issues. Let me know if which one you want me to pick up as I see you've started fixing a few (like the version check).

**Reply by bbqkees** [reply]
Would it be possible to add a notification in the web interface when someone tries to update from 3.8.2/devs to 3.9?
I can imagine a flood of emails when the older config does not migrate automatically.

**Reply by MichaelDvP** [reply]
Yes proddy has made such a message, works with the latest build:
<img width="600" height="253" alt="grafik" src="https://github.com/user-attachments/assets/d78ef535-4c44-499d-bca8-e490c3cbb6a6" />

**Reply by proddy** [reply]
please also review the page in the docs @bbqkees where we can keep updates and also reference the BBQKees Wiki if needed.

**Reply by bbqkees** [reply]
> Yes proddy has made such a message, works with the latest build: <img alt="grafik" width="600" height="253" src="https://private-user-images.githubusercontent.com/59284019/581989329-d78ef535-4c44-499d-bca8-e490c3cbb6a6.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzY4NjI2MjEsIm5iZiI6MTc3Njg2MjMyMSwicGF0aCI6Ii81OTI4NDAxOS81ODE5ODkzMjktZDc4ZWY1MzUtNGM0NC00OTlkLWJjYTgtZTQ5MGMzY2JiNmE2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA0MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNDIyVDEyNTIwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzMjZlYjJiMjVhNmQ4NGY3NDdiMTM5YmMwMWM1Nzg4NDE3NzJiNTAzN2FhOTQ2OGQ3MGNiYmQzNjNmZWU3OTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.hcCW2cEV6rgcWB3V4Pts1-9u9_4kbOmnKXmPL33Vld8">

Ok missed that one.

**Reply by bbqkees** [reply]
> please also review the page in the docs @bbqkees where we can keep updates and also reference the BBQKees Wiki if needed.

You mean this page?
https://docs.emsesp.org/FAQ#upgrading-the-firmware

Currently dev is on 3.8.2devX.
What is the plan/timeframe to proceed to 3.9.0?

**Reply by proddy** [reply]
> What is the plan/timeframe to proceed to 3.9.0?

soon, Michael and I are ironing out the final issues.
